### ISSUE TITLE
Question: get percentage of progress in dzsave

### ISSUE THREAD

**Issue by martinweihrauch** [kleisauke/net-vips#31]
Hi!
Thx so much for netVips!!
I would like to use dzsave. Is it possible to pass a C# Action<int> to get the percentage of the process? 
I would like to show the progress on the console or in a Windows desktop program.

Thanks

Martin

**Comment by kleisauke** [comment]
Hello @martinweihrauch,

You can just set the `VIPS_PROGRESS` environment variable for simple text feedback. I see this when setting `VIPS_PROGRESS`:
![progress](https://user-images.githubusercontent.com/12746591/54879953-b64e4600-4e3f-11e9-847a-4005742e6364.gif)

Unfortunately there is no option yet to get progress feedback in C#. I'll have to expand NetVips a bit for that. 

TODO List:
- [ ] Create P/Invoke layer of the [`g_signal_connect`](https://developer.gnome.org/gobject/stable/gobject-Signals.html#g-signal-connect) and [`vips_image_set_progress`](https://libvips.github.io/libvips/API/current/VipsImage.html#vips-image-set-progress`) functions.
- [ ] Marshal the `VipsProgress` and `GTimer` struct.
- [ ] Test the `preeval`, `eval`, and `posteval` signals. See [here](https://github.com/libvips/libvips/blob/master/libvips/iofuncs/image.c#L815-L822) for an example. 
- [ ] Create some convenience functions in `Image.cs` that wraps the above functions / structs. We could use the [`IProgress<T>` interface](https://docs.microsoft.com/en-us/dotnet/api/system.iprogress-1) for this.

**Comment by martinweihrauch** [comment]
Thx for the answer! I haven't worked with environment variables yet in C#. I am currently creating a WPF desktop application. 
Could you please give me a hint: If I call NetVips dzsave with my WPF application, how can I set an environment variable and "catch" the text?

**Comment by kleisauke** [comment]
I've just added support for attaching progress feedback. For example:
```csharp
// Load from a huge JPEG file
var image = Image.NewFromFile("huge.jpg", access: Enums.Access.Sequential);

var progress = new Progress<int>(value =>
{
    Console.Write($"\r{value}% complete");
});

// Attach progress feedback
image.SetProgress(progress);

// Save image pyramid
image.Dzsave("image-pyramid");
```

Showcase:
![showcase](https://user-images.githubusercontent.com/12746591/55077648-8baafa00-5098-11e9-8abf-7fd719d77046.gif)

If you want to test this, you can use the nightly version of NetVips. Add the `https://ci.appveyor.com/nuget/net-vips` feed in the `<packageSources>` section of your `NuGet.config`:
```
<packageSources>
  <add key="netvips-nightly" value="https://ci.appveyor.com/nuget/net-vips" />
</packageSources>
```

And update NetVips to 1.1.0. 

Thank for reporting this!

**Comment by martinweihrauch** [comment]
Thanks a lot.
I implemented everything, but experienced an error, which is about the assistant for managed debugging. Do you have any clue what that could be?


Assistent für verwaltetes Debuggen "CallbackOnCollectedDelegate" 
Für den von der Garbage Collection gesammelten Delegaten vom Typ "NetVips!NetVips.Internal.GCallback::Invoke" wurde ein Rückruf durchgeführt. Dies kann Anwendungsabstürze, Datenbeschädigung und -verlust zur Folge haben. Beim Übergeben von Delegaten an nicht verwalteten Code müssen die Delegaten von der verwalteten Anwendung beibehalten werden, bis sichergestellt ist, dass sie nie aufgerufen werden.

Ausnahme ausgelöst: "System.NullReferenceException" in NetVips.dll
Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt.

Das Programm "[15844] PathoZoomConverter.exe" wurde mit Code -1 (0xffffffff) beendet.

**Comment by martinweihrauch** [comment]
My code:
var image = Image.NewFromFile(wsiFilename, access: Enums.Access.Sequential);
       
                // Save image pyramid
                image.Dzsave(outFilename);

**Comment by martinweihrauch** [comment]
And there is a "System.NullReferenceException" in NetVips.dll after some percent of running dzsave...

**Comment by kleisauke** [comment]
@martinweihrauch I had forgotten to protect the delegate against premature garbage collection. It's fixed with commit https://github.com/kleisauke/net-vips/commit/a5ce84a96eb1e2479209bf18c72104cc2b968ddf.

This code example now runs without any problems:
```csharp
var im = Image.NewFromFile(Filename, access: Enums.Access.Sequential);

var progress = new Progress<int>(value =>
{
    Console.Write($"\r{value}% complete");
});
im.SetProgress(progress);

// Force garbage collection cycle to prove
// that the delegate doesn't get disposed
GC.Collect();
GC.WaitForPendingFinalizers();

// Save image pyramid
im.Dzsave("images/image-pyramid");
```

Thanks for reporting this dumbness!

**Comment by martinweihrauch** [comment]
:)))
This is not a dumbness, but small mishap in an otherwise great project :))
Do you ever plan to also accept a "cancel token"? But I believe, Vips doesnt support cancelling, correct?

**Comment by kleisauke** [comment]
If you want, I can adjust the method to this:
<details><summary>Example</summary>

```csharp
/// <summary>
/// Attach progress feedback, if required.
/// </summary>
/// <remarks>
/// You can use this function to update user-interfaces with 
/// progress feedback, for example:
/// <code language="lang-csharp">
/// var image = Image.NewFromFile("huge.jpg", access: Enums.Access.Sequential);
/// 
/// var progress = new Progress&lt;int&gt;(value =>
/// {
///     Console.Write($"\r{value}% complete");
/// });
/// image.SetProgress(progress);
/// 
/// image.Dzsave("image-pyramid");
/// </code>
/// </remarks>
/// <param name="progress">A provider for progress updates.</param>
/// <param name="token">Cancellation token to stop progress updates.</param>
public void SetProgress(IProgress<int> progress, CancellationToken token = default)
{
    VipsImage.SetProgress(this, progress == null ? 0 : 1);

    var lastPercent = 0;

    GCallback evalCallback = (imagePtr, progressPtr, userDataPtr) =>
    {
        // Stop reporting the progress if a cancellation
        // has been requested for this token
        if (token.IsCancellationRequested)
        {
            return;
        }

        var progressStruct = progressPtr.Dereference<VipsProgress.Struct>();
        if (progressStruct.Percent != lastPercent)
        {
            progress?.Report(progressStruct.Percent);
            lastPercent = progressStruct.Percent;
        }
    };

    // prevent it from being re-located or disposed of by the garbage collector
    _progress = GCHandle.Alloc(evalCallback);

    this.Connect(Internal.Enums.VipsEvaluation.Eval, evalCallback);
}
````
</details>

For example, this allows you to stop the progress reporting after 5 sec:
```csharp
var cts = new CancellationTokenSource();
cts.CancelAfter(5000);

var progress = new Progress<int>(value =>
{
    Console.Write($"\r{value}% complete");
});
im.SetProgress(progress, cts.Token);
```

Unfortunately, this will only stop the progress updates and not the image processing itself. libvips does support cancellation see:
https://github.com/libvips/libvips/blob/8e55732bcdcc9525a7ffbbde127298fb6a3523b8/libvips/iofuncs/image.c#L1639-L1659

But it's internal API. 

@jcupitt Is it possible to move the `vips_image_set_kill` function to the public API? Or is this function for internal use only? It seems that this method is only used in the deprecated `vips7compat.c`. Making this method public can help users to interrupt long-running image processing tasks (such as `vips_dzsave`).

**Comment by jcupitt** [comment]
Sure, let's make it public.

I'll just check exactly how nip2 does cancel.

**Comment by kleisauke** [comment]
I just checked how nip2 does this exactly. It seems that if the callback in `im_add_eval_callback` returns `-1` it will kill the image (with `vips_image_set_kill`), see:
https://github.com/libvips/nip2/blob/1b1b27f6b61096f0230b9d7db2d2d361fa4758ed/src/progress.c#L169
https://github.com/libvips/nip2/blob/1b1b27f6b61096f0230b9d7db2d2d361fa4758ed/src/imageinfo.c#L672-L674
https://github.com/libvips/nip2/blob/1b1b27f6b61096f0230b9d7db2d2d361fa4758ed/src/imageinfo.c#L728-L730
https://github.com/libvips/libvips/blob/4af242b59935a6f330c788d6dc0f4e90e1c1ad0f/libvips/deprecated/vips7compat.c#L409

I can create a PR to move the `vips_image_set_kill` function to the public API, if you like.

**Comment by jcupitt** [comment]
Ah yes, I think I remember now.

It's OK, I'll do it, it's a very simple change.

**Comment by kleisauke** [comment]
I've just released [NetVips 1.1.0-rc1](https://www.nuget.org/packages/NetVips/1.1.0-rc1) and [NetVips.Native 8.8.0-rc1](https://www.nuget.org/packages/NetVips.Native/8.8.0-rc1) (which contains the pre-compiled libvips 8.8.0-rc1 binaries for Linux, macOS and Windows) on NuGet. This release adds support for progress feedback (`image.SetProgress()`) and signal handling (`image.SignalConnect()`).

Support for associated images is also included. You can attach all associated images as metadata by setting the second parameter of `Openslideload`, for example:
```csharp
var image = Image.Openslideload("image.svs", attachAssociated: true)
```
<sup>(see relevant difference [here](https://github.com/kleisauke/net-vips/commit/b1585c9d341611cd884925a42da5bc5f47557a4f#diff-c43ebd8737d7640a93546cb4298216faR6661))</sup>

Or by doing:
```csharp
var image = Image.NewFromFile("image.svs", kwargs: new VOption
{
    {"attach_associated", true}
});
image.Dzsave("image-pyramid", container: "szi");
```

Note that you need the the libvips **`-all`** variant on Windows (instead than the `-web`) to include OpenSlide support. The pre-compiled libraries within the new NetVips.Native package don't support OpenSlide images.

If you would like to see what's new in libvips 8.8, please visit the the release notes of libvips:
https://libvips.github.io/libvips/2019/04/22/What's-new-in-8.8.html

**Comment by kleisauke** [comment]
NetVips v1.1.0 is now available.
